File: src/main/java/al132/alchemistry/recipes/ProbabilitySet.java
package al132.alchemistry.recipes;

import al132.alchemistry.utils.ListUtils;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Lists;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

import static al132.alchemistry.utils.StackUtils.areStacksEqualIgnoreQuantity;

public class ProbabilitySet {

    private List<ProbabilityGroup> set;
    public final boolean relativeProbability;
    public final int rolls;

    public ProbabilitySet(List<ProbabilityGroup> set) {
        this(set, true, 1);
    }

    public ProbabilitySet(List<ProbabilityGroup> set, boolean relativeProbability, int rolls) {
        this.set = set;
        this.relativeProbability = relativeProbability;
        this.rolls = rolls;
    }

    public List<ProbabilityGroup> getSet(){
        return set;
    }

    public List<ItemStack> toStackList() {
        ImmutableList.Builder<List<ItemStack>> builder = ImmutableList.builder();
        set.forEach(it -> builder.add(ImmutableList.copyOf(it.getOutputs())));
        return builder.build().stream()
                .flatMap(List::stream)
                .collect(Collectors.toList());
    }

    public double probabilityAtIndex(int index) {
        double sum = getTotalProbability();
        if (relativeProbability) return set.get(index).getProbability() / sum;
        else return set.get(index).getProbability();
    }

    private double getTotalProbability() {
        return set.stream().mapToDouble(ProbabilityGroup::getProbability).sum();
    }

    public NonNullList<ItemStack> calculateOutput() {
        NonNullList<ItemStack> temp = NonNullList.create();
        Random rando = new Random();
        for (int i = 1; i <= rolls; i++) {
            if (relativeProbability) {
                double totalProbability = getTotalProbability();
                double targetProbability = rando.nextDouble();
                double trackingProbability = 0.0;

                for (ProbabilityGroup component : set) {
                    trackingProbability += (component.getProbability() / totalProbability);
                    if (trackingProbability >= targetProbability) {
                        component.getOutputs().stream().filter(x -> !x.isEmpty()).forEach(x -> {
                            ItemStack stack = x.copy();
                            int index = ListUtils.indexOfFirst(temp, it -> areStacksEqualIgnoreQuantity(stack, it));
                            if (index != -1) temp.get(index).grow(stack.getCount());//stack.count)
                            else temp.add(stack);
                        });
                        break;
                    }
                }
            } else { //absolute probability
                for (ProbabilityGroup component : set) {
                    if (component.getProbability() >= rando.nextInt(101)) {
                        component.getOutputs().stream().filter(x -> !x.isEmpty()).forEach(x -> {
                            ItemStack stack = x.copy();
                            int index = ListUtils.indexOfFirst(temp, it -> areStacksEqualIgnoreQuantity(stack, it));
                            if (index != -1) temp.get(index).grow(stack.getCount());
                            else temp.add(stack);
                        });
                    }
                }
            }
        }
        return temp;
    }

    public static class Builder {
        private List<ProbabilityGroup> groups = new ArrayList<>();

        public Builder() {
        }


        public Builder addGroup(ProbabilityGroup group) {
            groups.add(group);
            return this;
        }

        public Builder addGroup(double probability, ItemStack... stacks) {
            groups.add(new ProbabilityGroup(Lists.newArrayList(stacks), probability));
            return this;
        }

        public ProbabilitySet build() {
            return new ProbabilitySet(this.groups);
        }
    }
}

File: src/main/kotlin/al132/alchemistry/recipes/ProbabilitySet.kt
package al132.alchemistry.recipes

import al132.alib.utils.extensions.toImmutable
import com.google.common.collect.ImmutableList
import net.minecraft.item.ItemStack
import java.util.*


data class ProbabilityGroup(private val _output: List<ItemStack>,
                            val probability: Double = 1.0) {

    val output: List<ItemStack>
        get(): List<ItemStack> = _output.toImmutable()
}


data class ProbabilitySet(private var _set: List<ProbabilityGroup>? = ArrayList<ProbabilityGroup>(),
                          val relativeProbability: Boolean = true,
                          val rolls: Int = 1) {

    val set: List<ProbabilityGroup>
        get(): List<ProbabilityGroup> = ImmutableList.copyOf(_set)


    fun toStackList(): List<ItemStack> {
        val temp = ImmutableList.Builder<ImmutableList<ItemStack>>()
        set.forEach { temp.add(ImmutableList.copyOf(it.output)) }
        return temp.build().flatten().toImmutable()
    }

    fun probabilityAtIndex(index: Int): Double {
        if (relativeProbability) return (set[index].probability.toDouble() / set.sumByDouble { it.probability })
        else return set[index].probability.toDouble()
    }


    fun calculateOutput(): List<ItemStack> {
        val temp = ArrayList<ItemStack>()
        val rando = Random()
        (1..rolls).forEach { _ ->
            if (relativeProbability) {
                val totalProbability = set.sumByDouble { it.probability }
                val targetProbability = rando.nextDouble()
                var trackingProbability = 0.0

                for (component in set) {
                    trackingProbability += (component.probability.toDouble() / totalProbability)
                    if (trackingProbability >= targetProbability) {
                        component.output.filterNot { it.isEmpty }.forEach { temp.add(it.copy()) }
                        break
                    }
                }
            } else { //absolute probability
                for (component in set) {
                    if (component.probability >= rando.nextInt(101)) {
                        component.output.filterNot { it.isEmpty }.forEach { temp.add(it.copy()) }
                    }
                }
            }
        }
        return temp
    }
}